In [18]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split

data = {
    'area': [50, 70, 80, 100, 120, 150, 40, 65, 90, 110],
    'rooms': [1, 2, 3, 3, 4, 5, 1, 2, 3, 4],
    'age': [5, 10, 2, 15, 20, 1, 30, 8, 12, 5],
    'price': [250, 350, 450, 480, 550, 700, 180, 310, 420, 530]
}
df = pd.DataFrame(data)

X = df[['area', 'rooms', 'age']]
y = df['price']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print('Размер обучающей выборки:', X_train.shape)

Размер обучающей выборки: (8, 3)


In [22]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

model = LinearRegression()
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

MAE = mean_absolute_error(y_test, y_pred)
MSE = mean_squared_error(y_test, y_pred)
RMSE = MSE**0.5
R2 = r2_score(y_test, y_pred)

print('MAE:', MAE)
print('MSE:', MSE)
print('RMSE:', RMSE)
print('R2 Score:', R2)
for feature, coef in zip(X.columns, model.coef_):
    print(f'Признак {feature}: влияние на цену {coef:.2f}')

MAE: 12.953727789068324
MSE: 181.5558565571787
RMSE: 13.474266457109222
R2 Score: 0.8517911375043439
Признак area: влияние на цену 2.68
Признак rooms: влияние на цену 45.43
Признак age: влияние на цену -1.55


In [23]:
print(y.min(), y.max())

180 700


Модель ошибается в среднем на ~13 единиц. Учитывая что диапазон целевой переменной составляет 180-700 (~520 единиц), ошибка составляет около 2.5% — отличный результат для линейной регрессии на прогнозирование цен на жилье.

MAE vs MSE:
Обе метрики показывают насколько ошибается модель, но по-разному:
- MAE (Mean Absolute Error) — среднее абсолютных ошибок. Все ошибки влияют одинаково.
- MSE (Mean Squared Error) — среднее квадратов ошибок. Большие ошибки штрафуются сильнее, так как возводятся в квадрат.
Если модель редко ошибается сильно — MAE и MSE будут близки (как в нашем случае). Если есть большие выбросы — MSE будет значительно выше.

R2 Score — показатель того, насколько модель справляется лучше, чем просто возвращать среднее значение.

Признаки показывают что каждая добавленная площадь в квм увеличивает цену жилья на 2.68 тыс. долларов. Также каждая добавленная комната добавляет к стоимости 45.43 тыс. долларов и возраст уменьшает цену дома на 1.55 тыс. долларов за каждый год. 

In [34]:
# реализация MSE вручную:
MyMSE = 1/len(y_test) * sum((y_test.values - y_pred)**2)
print("sklearn's MSE = ", MSE)
print("my MSE = ", MyMSE)
print("Same!" if MSE == MyMSE else "Error :(")

sklearn's MSE =  181.5558565571787
my MSE =  181.5558565571787
Same!
